### PR TITLE
Fix: X icon only reloads the page when arriving from an OAuth app

### PR DESCRIPTION
**Pull request by BryanEliDimas**
Fix https://github.com/openstax/business-intel/issues/968

### CODE REVIEW

File: app/controllers/newflow/login_signup_controller.rb [hunk 1]
```diff
@@ -326,6 +325,22 @@ def change_password
       end
     end
 
+    def exit_accounts
+      referrer_params = Addressable::URI.parse(request.referrer).query_values.to_h
+
+      if (return_to = referrer_params['r'])
```
Comment (steveburkett) [inline app/controllers/newflow/login_signup_controller.rb:331]: I would consider refactoring this to improve.    (1) not a great idea to assign variables inside of if statements.  (2) this method has high cyclometric complexity.  i encourage you to install and run flog on it https://blog.feabhas.com/2018/07/code-quality-cyclomatic-complexity/.   (3) does this method have unit tests covering all the if branches (one of the dangers of high cyclometric complexity, and bugs.

File: spec/features/newflow/student_login_flow_spec.rb [hunk 2]
```diff
@@ -139,6 +139,41 @@
         expect(page).to have_text('This is another banner.')
         screenshot!
       end
+
+      context 'when user clicks X icon' do
+        context "when user arrived from oauth app" do
+          let(:app) do
+            app = FactoryBot.create(:doorkeeper_application, skip_terms: true,
```
Comment (steveburkett) [inline spec/features/newflow/student_login_flow_spec.rb:146]: would be nice to see this broken up a bit into lets and befores.

Comment (BryanEliDimas) [reply spec/features/newflow/student_login_flow_spec.rb]: Fixed.

File: app/controllers/newflow/login_signup_controller.rb (hunk 1 above)
Comment (BryanEliDimas) [reply app/controllers/newflow/login_signup_controller.rb]: Re: 1. I agree that it's uncommon and should be avoided if necessary but I insist that doing so in this case is alright — especially after the refactoring I just did/pushed. Requesting re-review. :) 
2. Just glanced at the rather long article but I get the point, I know what you mean. 
3. I just improved all of my code immensely, IMO. I'm hoping that'll suffice. Requesting re-review. Thanks!

File: app/controllers/newflow/login_signup_controller.rb [hunk 3]
```diff
@@ -326,16 +327,16 @@ def change_password
     end
 
     def exit_accounts
-      referrer_params = Addressable::URI.parse(request.referrer).query_values.to_h
+      referrer_params = extract_params(request.referrer)
 
-      if (return_to = referrer_params['r'])
-        if Host.trusted?(return_to)
-          redirect_to(return_to)
+      if (r = referrer_params[:r])
```
Comment (steveburkett) [inline app/controllers/newflow/login_signup_controller.rb:332]: I would suggest a better name than r.  what is r?  it's good to write code thinking about how would a new dev look at this and instantly know what r is and what's going on

File: spec/controllers/newflow/login_signup_controller_spec.rb [hunk 4]
```diff
@@ -756,5 +756,63 @@ module Newflow
...
+        let(:target_url) { Faker::Internet.url }
+
+        context 'when Referer includes `r` param' do
+          subject do
+            "#{host}?r=#{target_url}"
+          end
+
+          context 'when `r` is trusted' do
+            before do
+              allow(Host).to receive(:trusted?).with(target_url).once.and_return(true)
```
Comment (steveburkett) [inline spec/controllers/newflow/login_signup_controller_spec.rb:776]: so much better! :)

File: app/controllers/newflow/login_signup_controller.rb (hunk 3 above)
Comment (BryanEliDimas) [reply app/controllers/newflow/login_signup_controller.rb:332]: 👍  I'm thinking of documenting this param in the https://github.com/openstax/accounts#special-parameters section of the README and/or adding a comment to the controller action `exit_accounts`.

**Pull request merged**